tests/e2e/test_spyre_cp_scheduler_steps.py

```diff
@@ -151,6 +151,161 @@ def test_prefill_tkv_too_big(model: ModelInfo, backend: str,
+    the tkv of the decode batch would be shifted beyond the max model length, 
+    we therefore have to wait with scheduling.
+
+    Configuration:
+        * max_num_seqs: 4
+        * number of prompts: 3
+            * 0: len = 20, max tokens = 17, step joining = 0
+            * 1: len = 80, max tokens = 17, step joining = 0
+            * 2: len = 16, max tokens = 50, step joining = 0
+    """
+
+    monkeypatch.setenv("VLLM_SPYRE_CP_INTERLEAVE_STEPS", "0")
+
+    seqs_max_tokens = [4, 3, 50]
+    prompts_lengths = [20, 80, 16]
+    steps_add_reqs = [0, 0, 0]
+
+    checked_steps = [
+        {
+            "step": 0,
+            "tkv": 0,
+            "waiting": ["0", "1", "2"],
+            "running": [],
+            "request_outputs": [],
+            "n_reserved_blocks": 0,
+            "n_used_blocks": 0
+        },
+        {
+            # Prefill sequence 0
+            "step": 1,
+            "tkv": 20,
+            "waiting": ["1", "2"],
+            "running": ["0"],
+            "request_outputs": ["0"],
+            "n_reserved_blocks": 1,
+            "n_used_blocks": 1
+        },
+        # tkv should be updated at the end of the last chunked prefill
+        # here we have only one chunk, so it will be updated directly
+        {
+            # Prefill sequence 1
+            "step": 2,
+            "tkv": 84,  # 64 (1 block padding) + 20 (prompt of seq 0) = 84
+            "waiting": ["2"],
+            "running": ["1", "0"],
+            "request_outputs": ["1"],
+            # 1 + 2 (prefill (2 block) + 3 decodes in the last block)
+            "n_reserved_blocks": 3,
+            "n_used_blocks": 3
+        },
+        # Here we cannot schedule sequence 2. Current tkv being in the second 
+        # block, the number of requested tokens can't fit in the remaining space 
+        # (64 (full block left padding) + 16 (prompt) + 50 (decode) = 130 > 128)
+        {
+            # Decode 1 of sequence 0
+            # Decode 1 of sequence 1
+            "step": 3,
+            "tkv": 85,
+            "waiting": ["2"],
+            "running": ["1", "0"],
+            "request_outputs": ["1", "0"],
+            "n_reserved_blocks": 3,
+            "n_used_blocks": 3
+        },
+        {
+            # Decode 2 of sequence 0
+            # Decode 2 of sequence 1
+            # Sequence 1 finishes
+            "step": 4,
+            "tkv": 86,
+            "waiting": ["2"],
+            "running": ["0"],
+            "request_outputs": ["1", "0"],
+            "finished_requests": ["1"],
+            "n_reserved_blocks": 3,
+            "n_used_blocks": 3
+        },
+        {
+            # Prefill sequence 2
+            # Left padding due to long prompt 1 can be removed, tkv is shifted
+            # back to the first block
+            #
+            # TODO: This step is failing because the tkv is still 86 right now even after the long prompt request 1 finishes
```

Comment:
@yannicks1 opinion here?

Reply:
thanks so much for investigating. 

> accept that we are suboptimal because the tkv our scheduler sees does not consider sequences that just left the batch in the last step.

this first solution makes sense to me, it is not dramatic to wait for one decode step to get the tkv updated. seems low priority to address this issue indeed. I'll update the tkv steps for that case